### PR TITLE
WIP Testing hypothesis

### LINKED ISSUE #107: Switching tests to Hypothesis
With [hypothesis](https://hypothesis.readthedocs.io/) providing a [pandas dataframe generator](https://hypothesis.readthedocs.io/en/latest/numpy.html#pandas), I'm wondering if we could generate more better battle-tested code by using hypothesis for property-based testing, instead of using our current hacky way of generating tests based on simple dataframes. 

One of Hypothesis' best traits is the ability to find edge cases that I myself was unable to find. I think this is worth a shot. Will tag this issue as appropriate.

### PR DESCRIPTION
**Pull request by ericmjl**
This PR resolves #107.

# PR Checklist

Please ensure that you have done the following:

1. [ ] PR in from a fork off your branch. Do not PR from <your_username>:master, but rather from <your_username>:<branch_name>.
2. [ ] If you're not on the contributors list, add yourself to `AUTHORS.rst`.

## Quick Check

To do a very quick check that everything is correct, follow these steps below:

- [ ] Run the command `make check` from pyjanitor's top-level directory. This will automatically run:
    - black formatting
    - pycodestyle checkin
    - running the test suite
    - docs build

## Code Changes

If you are adding code changes, please ensure the following:

- [ ] Ensure that you have added tests.
- [ ] Run all tests (`$ pytest .`) locally on your machine.
    - [ ] Check to ensure that test coverage covers the lines of code that you have added.
    - [ ] Ensure that all tests pass.

## Documentation Changes

If you are adding documentation changes, please ensure the following:

- [ ] Build the docs locally.
- [ ] View the docs to check that it renders correctly.

# PR Description

Please describe the changes proposed in the pull request:

- Refactors a bunch of tests that used a lot of repetitive code, taking advantage of pytest.mark.parametrize.
- Adds the use of Hypothesis' pandas dataframe generation capabilities to robustly generate dataframes for testing purposes. This may increase the test suite time, but it will also increase the quality of tests generated.

# Relevant Reviewers

Please tag maintainers to review.

- @zbarry 
- @szuckerman

### CODE REVIEW

File: environment-dev.yml [hunk 1]
```diff
@@ -6,7 +6,7 @@ channels:
 - ericmjl
 dependencies:
 - python=3.6
-- pandas
+- pandas >= 0.24.0
```
Comment (zbarry) [inline environment-dev.yml]: 👍

File: tests/test_functions.py [hunk 2]
```diff
@@ -159,78 +152,103 @@ def test_clean_names_original_columns(dataframe):
...
+    df2_dupes = df2.get_dupes()
     assert df2_dupes.shape == (0, 2)
 
 
-def test_encode_categorical():
-    df = pd.DataFrame()
-    df["class_label"] = ["test1", "test2", "test1", "test2"]
-    df["numbers"] = [1, 2, 3, 2]
-    df = encode_categorical(df, "class_label")
+def categoricaldf_strategy():
```
Comment (zbarry) [inline tests/test_functions.py]: `categoricaldf_strategy`

Kind of insignificant comment, but maybe rename to `categorical_df_strategy`? 🤔

**Comment by zbarry** [comment]
This Hypothesis library looks really cool. Don't understand how it works yet, but going to look into it for unit testing in the future.

**Comment by ericmjl** [comment]
The hypothesis library spares us from generating trivial examples for our tests, while also forcing us to specify tests by thinking about what properties the results should have, rather than what exact results should look like. 

I think for a beginner contributing, I wouldn't expect them to contribute a hypothesis-based test up-front, but rather provide one example-based test first, from which whoever of us is experienced with Hypothesis can generate a more robust test.

**Comment by ericmjl** [comment]
Let's take the conversation to the issue thread.
